File: packages/ui/db/entity/Area.ts
import { Boulder, Commentable, Crag, Photoable, Polygon } from "@/db";
import { cascadeManyToOne, cascadeOneToMany } from "@/db/cascadeOptions";
import BaseColumnSchemaPart from "@/db/entity/BaseColumnSchemaPart";
import { type IArea } from "models";
import { EntitySchema } from "typeorm";

export type Area = IArea & {
  polygon?: Polygon;
  crag?: Crag;
  boulders?: Boulder[];
  commentable?: Commentable;
  photoable?: Photoable;
};

const AreaSchema = new EntitySchema<Area>({
  name: "area",
  columns: {
    ...BaseColumnSchemaPart,
    name: {
      type: String,
    },
    description: {
      type: String,
      nullable: true,
    },
  },
  relations: {
    crag: {
      type: "many-to-one",
      target: "crag",
      ...cascadeManyToOne,
    },
    boulders: {
      type: "one-to-many",
      inverseSide: "area",
      target: "boulder",
      ...cascadeOneToMany,
    },
    polygon: {
      type: "one-to-one",
      target: "polygon",
      joinColumn: true,
      onDelete: "SET NULL",
      cascade: ["insert", "update"],
    },
    commentable: {
      type: "one-to-one",
      nullable: true,
      target: "commentable",
      onDelete: "SET NULL",
      joinColumn: true,
    },
    photoable: {
      type: "one-to-one",
      nullable: true,
      target: "photoable",
      onDelete: "SET NULL",
      joinColumn: true,
    },
  },
});

export default AreaSchema;

File: go-server/models/area.go
package models

type Area struct {
	ID          *int         `json:"id,omitzero"`
	Name        string       `json:"name"`
	Description *string      `json:"description,omitzero"`
	Crag        *Crag        `json:"crag,omitzero"`
	Polygon     *Polygon     `json:"polygon,omitzero"`
	Boulders    []Boulder    `json:"boulders,omitzero"`
	Commentable *Commentable `json:"commentable,omitzero"`
	Photoable   *Photoable   `json:"photoable,omitzero"`
}

func (a Area) IsZero() bool {
	return a.ID == nil && a.Name == "" && a.Description == nil && a.Crag == nil && a.Polygon == nil &&
		len(a.Boulders) == 0 && a.Commentable == nil && a.Photoable == nil
}
